File: ottoguide_web_app/frontend/src/mock/mockTelemetry.js
// Generador de telemetria simulada en el navegador (modo simulacion del front).
// Permite usar el panel sin backend ni robot. Espeja al mock del backend.
import { G1_JOINT_MAP } from '../data/jointMaps.js'

const ENTRIES = Object.entries(G1_JOINT_MAP).map(([idx, info]) => ({
  index: Number(idx),
  name: info.name,
  group: info.group,
}))
const NUM = ENTRIES.length

let start = performance.now()
const base = ENTRIES.map(() => (Math.random() * 30 - 15))

export function resetMock() {
  start = performance.now()
}

export function mockFrame() {
  const elapsed = (performance.now() - start) / 1000
  const t = elapsed
  const freq = 1.0

  const motors = ENTRIES.map((e, idx) => {
    const phase = idx * ((2 * Math.PI) / NUM)
    const tw = 2 * Math.PI * freq * t
    const angle = base[idx] + 10 * Math.sin(tw + phase)
    const vel = 10 * 2 * Math.PI * freq * Math.cos(tw + phase)
    const torque = 5 * Math.sin(tw + phase) + (Math.random() * 2 - 1)
    // La rodilla izquierda (idx 3) calienta un poco mas, para mostrar colores/alertas.
    const temp =
      idx === 3
        ? Math.round(44 + 6 * Math.sin(elapsed / 8) + (Math.random() * 2 - 1))
        : Math.round(30 + Math.random() * 11)
    return {
      index: e.index,
      name: e.name,
      group: e.group,
      q_rad: +(angle * Math.PI / 180).toFixed(4),
      q_deg: +angle.toFixed(2),
      dq: +vel.toFixed(4),
      ddq: 0,
      tau_est: +torque.toFixed(4),
      temperature: temp,
    }
  })

  const v = +(27.5 + (Math.random() * 0.6 - 0.3)).toFixed(2)
  const a = +(3.5 + (Math.random() * 1.0 - 0.5)).toFixed(2)
  const r = (n) => +(Math.random() * 2 * n - n).toFixed(4)
  const cell = (b) => b + Math.round(Math.random() * 8 - 4)

  return {
    robot: 'G1',
    timestamp: Date.now() / 1000,
    power_v: v,
    power_a: a,
    motors,
    imu: {
      quaternion: [1, 0, 0, 0],
      gyroscope: [r(0.05), r(0.05), r(0.05)],
      accelerometer: [r(0.3), r(0.3), +(9.8 + r(0.05)).toFixed(4)],
      rpy_deg: [+(r(3)).toFixed(2), +(r(3)).toFixed(2), +(r(10)).toFixed(2)],
    },
    foot_force: [0, 1, 2, 3].map(() => Math.round(15 + Math.random() * 40)),
    bms: {
      soc: Math.max(0, 75 - Math.floor(elapsed / 60)),
      cycle: 20,
      current: -3500,
      status: 0,
      mcu_ntc: [30, 31],
      bq_ntc: [29, 30],
      cell_vol: [cell(3520), cell(3518), cell(3525), cell(3530), cell(3522), cell(3519)],
    },
  }
}

File: ottoguide_web_app/backend/app/mock/mock_telemetry.py
"""
Generador de telemetria simulada para el robot G1 (29 motores).
Reproduce el tipo de datos que en produccion vendrian del lowstate por DDS:
energia, motores (angulo/velocidad/torque/temperatura), IMU, fuerza en patas y bateria.
"""
import math
import random
import time

# Mapa de motores del G1 (indice -> nombre, grupo)
G1_JOINT_MAP = [
    ("L_Hip_Pitch", "Pierna Izquierda"), ("L_Hip_Roll", "Pierna Izquierda"),
    ("L_Hip_Yaw", "Pierna Izquierda"), ("L_Knee", "Pierna Izquierda"),
    ("L_Ankle_Pitch", "Pierna Izquierda"), ("L_Ankle_Roll", "Pierna Izquierda"),
    ("R_Hip_Pitch", "Pierna Derecha"), ("R_Hip_Roll", "Pierna Derecha"),
    ("R_Hip_Yaw", "Pierna Derecha"), ("R_Knee", "Pierna Derecha"),
    ("R_Ankle_Pitch", "Pierna Derecha"), ("R_Ankle_Roll", "Pierna Derecha"),
    ("Waist_Yaw", "Cintura"), ("Waist_Roll", "Cintura"), ("Waist_Pitch", "Cintura"),
    ("L_Shoulder_Pitch", "Brazo Izquierdo"), ("L_Shoulder_Roll", "Brazo Izquierdo"),
    ("L_Shoulder_Yaw", "Brazo Izquierdo"), ("L_Elbow", "Brazo Izquierdo"),
    ("L_Wrist_Roll", "Mano Izquierda"), ("L_Wrist_Pitch", "Mano Izquierda"),
    ("L_Wrist_Yaw", "Mano Izquierda"),
    ("R_Shoulder_Pitch", "Brazo Derecho"), ("R_Shoulder_Roll", "Brazo Derecho"),
    ("R_Shoulder_Yaw", "Brazo Derecho"), ("R_Elbow", "Brazo Derecho"),
    ("R_Wrist_Roll", "Mano Derecha"), ("R_Wrist_Pitch", "Mano Derecha"),
    ("R_Wrist_Yaw", "Mano Derecha"),
]
NUM_MOTORS = len(G1_JOINT_MAP)


class MockTelemetry:
    def __init__(self):
        self._start = time.time()
        # angulo base fijo por motor
        self._base = [random.uniform(-15, 15) for _ in range(NUM_MOTORS)]

    def frame(self) -> dict:
        elapsed = time.time() - self._start
        t = elapsed
        freq = 1.0
        motors = []
        for idx, (name, group) in enumerate(G1_JOINT_MAP):
            base = self._base[idx]
            phase = idx * (2 * math.pi / NUM_MOTORS)
            tw = 2 * math.pi * freq * t
            angle = base + 10 * math.sin(tw + phase)
            vel = 10 * 2 * math.pi * freq * math.cos(tw + phase)
            torque = 5 * math.sin(tw + phase) + random.uniform(-1, 1)

            # La rodilla izquierda (idx 3) oscila un poco mas caliente para mostrar
            # los colores de temperatura y el banner de aviso de vez en cuando.
            if idx == 3:
                temp = int(44 + 6 * math.sin(elapsed / 8.0) + random.uniform(-1, 1))
            else:
                temp = random.randint(30, 41)

            motors.append({
                "index": idx,
                "name": name,
                "group": group,
                "q_rad": round(math.radians(angle), 4),
                "q_deg": round(angle, 2),
                "dq": round(vel, 4),
                "ddq": 0.0,
                "tau_est": round(torque, 4),
                "temperature": temp,
            })

        v = round(27.5 + random.uniform(-0.3, 0.3), 2)
        a = round(3.5 + random.uniform(-0.5, 0.5), 2)

        return {
            "robot": "G1",
            "timestamp": time.time(),
            "power_v": v,
            "power_a": a,
            "motors": motors,
            "imu": {
                "quaternion": [1, 0, 0, 0],
                "gyroscope": [round(random.uniform(-0.05, 0.05), 4) for _ in range(3)],
                "accelerometer": [
                    round(random.uniform(-0.3, 0.3), 4),
                    round(random.uniform(-0.3, 0.3), 4),
                    round(9.8 + random.uniform(-0.05, 0.05), 4),
                ],
                "rpy_deg": [
                    round(random.uniform(-3, 3), 2),
                    round(random.uniform(-3, 3), 2),
                    round(random.uniform(-10, 10), 2),
                ],
            },
            "foot_force": [random.randint(15, 55) for _ in range(4)],
            "bms": {
                "soc": max(0, 75 - int(elapsed / 60)),
                "cycle": 20,
                "current": -3500,
                "status": 0,
                "mcu_ntc": [30, 31],
                "bq_ntc": [29, 30],
                "cell_vol": [
                    3520 + random.randint(-4, 4), 3518 + random.randint(-4, 4),
                    3525 + random.randint(-4, 4), 3530 + random.randint(-4, 4),
                    3522 + random.randint(-4, 4), 3519 + random.randint(-4, 4),
                ],
            },
        }


mock = MockTelemetry()
